integration/sandbox_clean_remove_windows_test.go

```diff
@@ -147,3 +194,119 @@ func TestSandboxRemoveWithoutIPLeakage(t *testing.T) {
+	}
+}
+
+func stopContainer(ctx context.Context, t *testing.T, client runtime.RuntimeServiceClient, containerID string) {
+	_, err := client.StopContainer(ctx, &runtime.StopContainerRequest{
+		ContainerId: containerID,
+		Timeout:     0,
+	})
+	if err != nil {
+		t.Fatalf("failed StopContainer request for container: %s, with: %v", containerID, err)
+	}
+}
+
+func startContainer(ctx context.Context, t *testing.T, client runtime.RuntimeServiceClient, containerID string) {
+	_, err := client.StartContainer(ctx, &runtime.StartContainerRequest{
+		ContainerId: containerID,
+	})
+	if err != nil {
+		t.Fatalf("failed StartContainer request for container: %s, with: %v", containerID, err)
+	}
+}
+
+func removeContainer(ctx context.Context, t *testing.T, client runtime.RuntimeServiceClient, containerID string) {
+	_, err := client.RemoveContainer(ctx, &runtime.RemoveContainerRequest{
+		ContainerId: containerID,
+	})
+	if err != nil {
+		t.Fatalf("failed RemoveContainer request for container: %s, with: %v", containerID, err)
+	}
+}
+
+// This test checks if create/stop and remove pods and containers work as expected
+func Test_CreateContainer_Test(t *testing.T) {
+	testImage := getTestImage()
+	if testImage == "" {
+		t.Skip("skipping test")
+	}
+	client, err := RawRuntimeClient()
+	require.NoError(t, err)
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	t.Log("Create a pod sandbox")
+	sbConfig := &runtime.PodSandboxConfig{
+		Metadata: &runtime.PodSandboxMetadata{
+			Name: t.Name(),
+		},
+	}
+	sandboxRequest := &runtime.RunPodSandboxRequest{
+		Config:         sbConfig,
+		RuntimeHandler: "runhcs-wcow-process",
+	}
+	sandBoxResponse, err := client.RunPodSandbox(ctx, sandboxRequest)
+	if err != nil {
+		t.Fatalf("failed RunPodSandbox request with: %v", err)
+	}
+	// Make sure the sandbox is cleaned up.
+	defer removePodSandbox(ctx, t, client, sandBoxResponse.PodSandboxId)
+	defer stopPodSandbox(ctx, t, client, sandBoxResponse.PodSandboxId)
+
+	EnsureImageExists(t, testImage)
+
+	t.Log("Create a container")
+	createCtrRequest := &runtime.CreateContainerRequest{
+		Config: &runtime.ContainerConfig{
+			Metadata: &runtime.ContainerMetadata{
+				Name: t.Name() + "-Container",
+			},
+
+			Image: &runtime.ImageSpec{
+				Image: testImage,
+			},
+			Command: []string{
+				"cmd",
+				"/c",
+				"ping",
+				"-t",
+				"127.0.0.1",
+			},
+			Windows: &runtime.WindowsContainerConfig{
+				Resources: &runtime.WindowsContainerResources{
+					CpuShares: 500,
+				},
+			},
+		},
+		PodSandboxId:  sandBoxResponse.PodSandboxId,
+		SandboxConfig: sandboxRequest.Config,
+	}
+
+	createCtrResponse, err := client.CreateContainer(ctx, createCtrRequest)
+	if err != nil {
+		t.Fatalf("failed CreateContainer in sandbox: %s, with: %v", sandBoxResponse.PodSandboxId, err)
+	}
+
+	// Make sure the container is cleaned up.
+	defer removeContainer(ctx, t, client, createCtrResponse.ContainerId)
```

Comment:
Hi @samuelkarp , I noticed that changing defer to t.Cleanup() is causing the test to fail with "rpc error: code = Canceled desc = context canceled" while calling removeContainer(). I noticed that when Cleanup() is calling the helper function, the test has already finished/exited and the context is cancelled when the helper function is called. But in defer the test has still not completed so the context is still valid and removeContainer() completes successfully. So I changed it back to a defer. Please let me know if you have any concerns. Thanks!

Reply:
I think that's because you explicitly have `defer cancel()` on line 202.  A `context.Background()` wouldn't otherwise be canceled when the test exited.

If you want to cancel the context, do that in a `t.Cleanup` as well and it should execute in the right order.

Reply:
I think either way (changing the `defer cancel()` to use `t.Cleanup`, or changing the `t.Cleanup` calls back to `defer`) will ultimately be fine. Both `defer` and `t.Cleanup` should be invoked on panic. The only real difference I'm aware of is `t.Cleanup` can be registered from a helper function, which isn't relevant here since we're calling from the main test function.